proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,556 @@
+
+::
+
+  printf
+    [__i|
+    instance Aeson.FromJSON %s where
+      parseJSON =
+        Aeson.withText "%s" $ \s ->
+          either Aeson.parseFail pure $
+            parsePrinterOptType (Text.unpack s)
+
+    instance PrinterOptsFieldType %s where
+      parsePrinterOptType s =
+        case s of
+    %s
+          _ ->
+            Left . unlines $
+              [ "unknown value: " <> show s
+              , "Valid values are: %s"
+              ]
+    |]
+    fieldTypeName
+    fieldTypeName
+    fieldTypeName
+    ( unlines_
+        [ printf "      \"%s\" -> Right %s" val con
+        | (con, val) <- enumOptions
+        ]
+    )
+    (renderEnumOptions enumOptions)
+
+With multiline strings:
+
+::
+
+  printf
+    """
+    instance Aeson.FromJSON %s where
+      parseJSON =
+        Aeson.withText "%s" $ \\s ->
+          either Aeson.parseFail pure $
+            parsePrinterOptType (Text.unpack s)
+
+    instance PrinterOptsFieldType %s where
+      parsePrinterOptType s =
+        case s of
+    %s
+          _ ->
+            Left . unlines $
+              [ "unknown value: " <> show s
+              , "Valid values are: %s"
+              ]
+    """
+    fieldTypeName
+    fieldTypeName
+    fieldTypeName
+    ( unlines_
+        [ printf "      \"%s\" -> Right %s" val con
+        | (con, val) <- enumOptions
+        ]
+    )
+    (renderEnumOptions enumOptions)
+
+Effect and Interactions
+-----------------------
+
+A multiline string should be the same as a normal string after parsing, so ``OverloadedStrings`` and any other language features should work as usual.
+
+Should not break existing code, unless someone is actually using ``"""a"""`` to mean ``"" "a" ""``.
+
+Costs and Drawbacks
+-------------------
+
+Since this only affects lexing and parsing, I expect development and maintenance costs to be low. This feature is common in other languages, so there shouldn't be any learning curve for new developers coming from another language. If anything, the auto-stripping of leading whitespace might be a source of confusion, but a one-line explanation should be sufficient.
+
+Alternatives
+------------
+
+* Status quo, e.g. using ``unlines``
+
+  * As mentioned in the Motivation, it's not great ergonomics, but it works.
+
+* Third party libraries, using quasiquoters
+
+  * Template Haskell is overkill for this
+
+* No stripping of leading whitespace
+
+  * This probably comes from one of two concerns: more complex implementation, conceptually adds automagic. It does make the implementation a bit harder, but this is a small enough change that I don't think it makes the overall proposal much harder to implement. While it does add a bit more magic behind the scenes, I think the rule is simple enough (no more complex than do-block indentation rules) and the use-case common enough (I can't think of a single use-case that would want the indentation to be part of the string) that it warrants the bump in ergonomics.
```

Comment:
One way of addressing this concern is to require an explicit delimiter for the leading margin, like in the "indent each line" example. Have you considered limiting margin stripping to cases with explicit delimiters?

Reply:
Sure, like the [Kotlin approach](https://kotlinlang.org/docs/strings.html#multiline-strings):

```kotlin
// "      a\n      b\n      c"
val text = """
      a
      b
      c
      """

// "a\nb\nc"
val text = """
      |a
      |b
      |c
      """.trimMargin()

// "a\nb\nc"
val text = """
      #a
      #b
      #c
      """.trimMargin("#")
```

Given this was half a year ago, I don't remember if I considered this or not :sweat_smile: Some comments:
* Generally not a fan of special casing a specific character to delimit leading margin. Plus, wanting to actually start a line with that character would require escaping it
* So if we don't want to special case a specific character, we'd have to introduce a `trimMargin` function to Prelude (and probably a `trimMarginWith` function to customize the character)
    * In addition to wanting to avoid adding a function for this feature, the downside of this means the margin trimming would happen at runtime instead of compile time
* A priori, I believe margin stripping is much more common than not-margin stripping. Especially given Haskell's indentation-based layout, I think devs would expect reindenting code to not change behavior, which is true with always-margin-stripping and not with never-margin-stripping.

I'll put all of this in the proposal. Thanks for the insight!

Reply:
Presumably, a `trimMargin` function would also not work with `OverloadedStrings`.